### PR TITLE
Add check public

### PR DESCRIPTION
**Pull request by jcollado**
In this PR `addCheck` is made public to be able to add custom validators and an example is added to `README`.

I'd say that the API is confusing because custom validators cannot be added to `validimir` and make it available for every new validation function. I'd prefer something along these lines:

```js
var v = require('validimir')
v.addCheck('myCheck', function() { ... });
var fn = v().string().len(...).myCheck();
```

Please let me know your feedback/comments.

Related #2

### CODE REVIEW

**Comment by juliangruber** [comment]
The problem with your example is that the check would be added globally, i.e. every module requiring `validimir` would have `myCheck` and thus modules could overwrite each others' checks.

What do you think about something like this?

```js
var v = require('validimir')
function myCheck() { ... });
var fn = v().string().len(...).custom(myCheck);
```

As I said before I have no objection to just adding a simple ip address check to validimir master, until we figure out a sane extension pattern.

I could also imagine something like this:

```js
var v = require('validimir')(); // get your own instance
v.addCheck('myCheck', function(){ ... });
var fn = v().string().len(...).myCheck();
```

**Comment by jcollado** [comment]
The first example looks good to me. Let me see if I can update this PR to behave like that.

Regarding the second example, note that it won't work because there are two different `v()` calls and the new check added would be lost in the second call.

**Comment by juliangruber** [comment]
the 2nd example could be made to work, it's just a breaking change to the current ap

**Comment by jcollado** [comment]
Yes, but that would cause checks to be shared globally, which you said to prefer not to happen.

In the example in the `README` in this PR, checks are not shared; but, as I said, I don't like much the API because the style used in all examples is chaining calls except for the check addition. I like the `custom` method because it preserves that style.

**Comment by juliangruber** [comment]
note the

```js
var v = require('validimir')();
```

instead of

```js
var v = require('validimir');
```

**Comment by jcollado** [comment]
I didn't see that; but still it's confusing in my opinion because `v()` is used later which would be equivalent to `require('validimir')()()` and that's supposed to work the same way as `require('validimir')()`.

**Comment by jcollado** [comment]
@juliangruber I've updated the code to use the `custom` proposal. Please let me know your thoughts. Thanks.

File: README.md [hunk 1]
```diff
@@ -103,6 +103,46 @@ npm install validimir
...
+  - `expected`: The expected value, if any, depending on the check
+
+  Example:
+
+```js
+var v = require('validimir');
+var isIP = require('validator').isIP
+var check = function(value) {
+  if (!isIP(value)) {
+    return {}
```
Comment (juliangruber) [inline README.md]: double return

File: index.js [hunk 2]
```diff
@@ -28,6 +28,13 @@ module.exports = function V(){
     };
   };
 
+
+  // Add custom validator
+  v.custom = function(fn) {
+    checks.push(fn);
+    return v;
+  };
```
Comment (juliangruber) [inline index.js]: simpler, using existing apis:

```js
addCheck('custom', function(fn){
  return fn;
});
```

File: README.md (hunk 1 above)
Comment (jcollado) [inline README.md]: Oops, that was there for testing. Removed.

**Comment by jcollado** [comment]
@juliangruber All your comments have been addressed. Please let me know if you have any other comment.

File: README.md [hunk 3]
```diff
@@ -103,6 +103,45 @@ npm install validimir
...
+    return {
+      value: value,
+      operator: 'ip',
+      message: 'Expected a valid IP address'
+    }
+  }
+}
+var fn = v().custom(check)
+
+fn('not an ip')
```
Comment (juliangruber) [inline README.md]: we can simplify the comment below if this read `fn('not an ip').errors`

File: README.md [hunk 4]
```diff
@@ -103,6 +103,45 @@ npm install validimir
...
+  - `operator`: The name of the operator that failed
+  - `message`: Descriptive error messsage
+  - `expected`: The expected value, if any, depending on the check
+
+  Example:
+
+```js
+var v = require('validimir');
+var isIP = require('validator').isIP
+var check = function(value) {
```
Comment (juliangruber) [inline README.md]: let's rename this to `checkIP` to make it more real-world

File: README.md [hunk 5]
```diff
@@ -103,6 +103,45 @@ npm install validimir
...
+  - `value`: The value that didn't pass the validation
+  - `operator`: The name of the operator that failed
+  - `message`: Descriptive error messsage
+  - `expected`: The expected value, if any, depending on the check
+
+  Example:
+
+```js
+var v = require('validimir');
+var isIP = require('validator').isIP
```
Comment (juliangruber) [inline README.md]: semicolon, also some missing below

File: index.js [hunk 6]
```diff
@@ -28,6 +28,12 @@ module.exports = function V(){
     };
   };
 
+
+  // Add custom validator
```
Comment (juliangruber) [inline index.js]: this project is comment free so far, let's stick to the style and not introduce some

File: README.md (hunk 4 above)
Comment (jcollado) [inline README.md]: Renamed in 4a2d3b3

File: README.md (hunk 3 above)
Comment (jcollado) [inline README.md]: Updated in b34f2f1

File: README.md (hunk 5 above)
Comment (jcollado) [inline README.md]: @juliangruber I'm not sure what's the policy in this project regarding semicolons, could you elaborate on this?

File: index.js (hunk 6 above)
Comment (jcollado) [inline index.js]: Removed in 15577a5

File: README.md (hunk 5 above)
Comment (juliangruber) [inline README.md]: if you look at https://github.com/juliangruber/validimir/blob/master/index.js and https://github.com/juliangruber/validimir/blob/master/README.md, there's consistent use of semicolons - unless i missed one

Comment (jcollado) [inline README.md]: @juliangruber Ok, so I added missing semicolons to the example in 383c034 and a couple more that I think you missed in 873f98b

**Comment by jcollado** [comment]
@juliangruber All your comments have been addressed now. Please let me know if you have some other comment/suggestion. Thanks.

File: test/test.js [hunk 7]
```diff
@@ -507,3 +507,12 @@ test('putin', function(t) {
   });
 });
 
+
+test('custom', function(t) {
```
Comment (juliangruber) [inline test/test.js]: missing semicolons :P

Comment (jcollado) [inline test/test.js]: @juliangruber Fixed in 53de595

**Comment by jcollado** [comment]
@juliangruber Added a configuration file for `eslint` that should be helpful to address style problems automatically instead of in a code review.

Please let me know if you have some other comment. Thanks.

**Comment by juliangruber** [comment]
:egg:cellent! published to on as 1.5.0

**Pull request merged**